### PR TITLE
[WIP - comments requested] query based interface

### PR DESCRIPTION
**Pull request by isaacsas**
This is another attempt to use a more delayed parsing approach. My goal is to allow for the exact same set of fields as the current macro generates, but with most of the function-based fields generated the first time the corresponding problem is created. 

Some notes
1. All tests pass but one;`mass_act_jump_tests.jl` at line 83 consistently fails, giving me an error about the world age. I'm quite stumped what the issue is here, so if anyone has suggestions please let me know. It seems to be something related to the `jumps` not being defined in a newer world than `DiffEqJump`'s `Direct` SSA.
2. I'm not sure if using `Unions` in the `reaction_network` type could lead to performance issues. We'll need to investigate this.

### CODE REVIEW

**Comment by TorkelE** [comment]
Comments:

I agree that calculating things when they first are required should be a better idea (In ODE, SDE, Jump Problems etc.). I do think we should provide a function like `complete_reaction_network!(rn::::DiffEqBase.AbstractReactionNetwork)`. Creating a reaction network and then calling that function should have similar effect to right now.

I am not sure whenever I think storing all the required information in a dictionary right now is a good idea. Maybe at some point we start to add a lot of fields to the structure and this would be a good idea. But I'd argue that we should wait until then to make this change. This change just adds a lot of notation when one wants to access fields in the structure, e.g. `rn.f` becomes `rn.properties[:f]`. This while there (right now) is no real gain of the change. Also, the change is breaking and risks messing up quite a lot of code.

A consequence of this change is that we have to use `eval` to generate the various functions, instead of the macro taking care of the evaluation. My knowledge of meta-programming is restricted. I know that `eval` is potentially dangerous to use (something with always being in global scope, I think), however, I think that this is the right circumstances to use it (and that it shouldn't have a to bad effect). But it would be good to have the opinion of someone which understands it to what effect it will have.

I am not sure I understand the new way the system works, taking `ODEProblem` as example. Currently we do not even have a problem (I think the type inherits from directly from functions). However, I would write it all something like this:
```julia
DiffEqBase.ODEProblem(rn::DiffEqBase.AbstractReactionNetwork, kwargs...) =
(rn.f==nothing)&&gen_f(rn)


```
    SDEProblem(rn, rn.g::Function, u0, args...;noise_rate_prototype=rn.p_matrix, kwargs...)

**Comment by TorkelE** [comment]
(Some random weirdness when I am writing a comment and manage to press the wrong button with a half written message and at the same time close the stuff, very much not intended...)

**Comment by TorkelE** [comment]
Comments:

I agree that calculating things when they first are required should be a better idea (In ODE, SDE, Jump Problems etc.). I do think we should provide a function like `complete_reaction_network!(rn::::DiffEqBase.AbstractReactionNetwork)`. Creating a reaction network and then calling that function should have similar effect to right now.

I am not sure whenever I think storing all the required information in a dictionary is a good idea right now. At some point we might start to add a lot of fields to the structure and then this would be useful. But I'd argue that we should wait until when/if that happens to make this change. This do add a lot of notation when one wants to access fields in the structure, e.g. `rn.f` becomes `rn.properties[:f]`, while providing new direct advantage right now. Also, the change is breaking and risks messing up quite a lot of code.

A consequence of this approach is that we have to use `eval` to generate the various functions, instead of the macro taking care of the evaluation. My knowledge of meta-programming is restricted. I know that `eval` is potentially dangerous to use (something with always being in global scope, I think), however, I also think that this is the right circumstances to use it (and that it shouldn't have a to bad effect). But it would be good to have the opinion of someone which understands `eval` and metaprogamming better to say what effect this will have.

Finally I am not sure I understandhow the new system works, taking `ODEProblem` as example: Currently we do not even have a problem for it (I think the type inherits directly from functions). However, it would look something like this:
```julia
DiffEqBase.ODEProblem(rn::DiffEqBase.AbstractReactionNetwork, kwargs...) =
ODEProblem(rn.f, kwargs...)
```
with the desire to do the calculations later I would change this to:
```julia
DiffEqBase.ODEProblem(rn::DiffEqBase.AbstractReactionNetwork, kwargs...) =
(rn.f==nothing)&&gen_f(rn)
ODEProblem(rn.f, kwargs...)
```
where `gen_f(rn)` is pretty much `f_expr = get_f(reactions, reactants)` and `f = make_func(f_expr, reactants, parameters)` from the `coordinate` function. Now I am not sure if this would mean problem that we are changing a `struct` after it is created. However, ignoring that, this would seem to me like the simple solution. Now we have two different `ODEProblem` function created, a `gen_odefun_inplace!` and a `get_odefun!` function, a `ODEReactionNetwork` struct and a `ODEReactionNetwork` function to generate it with (which only get called by the `SDEReactionNetwork` function which do not get called at all). There is probably a good reason for this, but it would be good to hear the thoughts behind it to help understanding (because reading other peoples code is hard, especially here on git were I cannot get an overview of the files but just get a long bloated document with all the changes).

**Comment by ChrisRackauckas** [comment]
One way to keep the current fields is to type them in a way that isn't strict (leave them as Any), constructed at first being nothing and then added later.

**Comment by TorkelE** [comment]
It still feels like we are doing much more than what should actually be needed. If we only want to rearrange where a few things are computed there is no reason we would now have to declare a `ODEReactionNetwork` structure.

Ideally I would simply keep all syntax the same and change the problem definition to
```julia
DiffEqBase.ODEProblem(rn::DiffEqBase.AbstractReactionNetwork, kwargs...) =
(rn.f==nothing)&&gen_f(rn)
ODEProblem(rn.f, kwargs...)
```
this would change nothing more than what we actually would like to accomplish (delaying some computations until the problem is defined). But it would require us to remake reaction_network into a `mutable struct`, this is bad right? Is there maybe some way to trick around this?

Alternatively we could allow for a separate macro which creates a barebone reaction network. One can then do operations on this one and if one wants to start simulations a conversion function could be provided (admittedly this feels like a less than ideal solution).

Basically, it feels like what we want to do should be doable without introducing breaking changes.

I will have another look, as well as trying to figure out what causes the world age problem and how to avoid it, next week.

**Comment by isaacsas** [comment]
@ChrisRackauckas @TorkelE Here's another version that only uses a single struct generated by the `reaction_network`, but makes the deferred fields unions with `Nothing`. I think this is more along the lines of what you both were suggesting, and shouldn't break any existing interfaces.

As in all my previous versions, the one jump test at line 83 in `mass_act_jump_tests.jl` is causing a world age error to appear when the `Direct` method simulation is running. This is related to the jump functions being defined in a newer world than the `Direct` method.

I'm also still not sure if there are performance implications to generating the various functions here through using `eval` instead of a macro at the highest calling level. This is something I can explore further once the world age bug gets resolved.

File: src/maketype.jl [hunk 1]
```diff
@@ -1,72 +1,127 @@
...
 
     # Make the type instance using the default constructor
     typeex,constructorex
 end
+
+function gen_ode!(rn::DiffEqBase.AbstractReactionNetwork)
+    @unpack reactions, syms_to_ints, params_to_ints = rn
+
+    f_expr        = get_f(reactions, syms_to_ints)
+    rn.f          = eval(make_func(f_expr, syms_to_ints, params_to_ints))
```
Comment (ChrisRackauckas) [inline src/maketype.jl]: evaling a function definition like this in a function seems like it's always going to be prone to world age problems. I am wondering if this strategy only could work well in global scope macros.

Comment (isaacsas) [reply src/maketype.jl]: I've been wondering the same. That said, we don't seem to see any problems in tests when we just define functions (ie the ODE and SDE function fields). Where the problem seems to occur is with the types that are instantiated and wrap functions, i.e. `ConstantRateJumps`. 

I tried making a macro, but 
1. Couldn't figure out how to get it to actually construct the jumps without calling `eval` within it too.
2. Even getting the jumps to be defined in the top level by using `Base.eval(@__MODULE__,jumps)` I still got world age errors.

One other note, using function wrappers for `ConstantRateJumps`, i.e. `DirectFW`, gets rid of the test failure.

Comment (isaacsas) [reply src/maketype.jl]: Also, correct me if I'm wrong, but doesn't `ModelingToolkit` do something similar in creating `ODEFunctions` from a `DiffEqSystem`?

Comment (ChrisRackauckas) [reply src/maketype.jl]: ModelingToolkit.jl does have an eval form for using it directly, but it's generally made to be used as the IR inside of a macro. I.e., it's made so that macros like the reaction DSL to do its internal calculations, and that wouldn't need eval.

Comment (isaacsas) [reply src/maketype.jl]: Got it. It seems like it may have reached the point that I should try to cook up a MWE and see if anyone has suggestions on Discourse.

File: test/mass_act_jump_tests.jl [hunk 2]
```diff
@@ -80,6 +80,7 @@ u0           = [0,0,0,0]
 u0[something(findfirst(isequal(:DNA),rs.syms),0)] = 1
 expected_avg = 5.926553750000000e+02
 rates = [.5, (20*log(2.)/120.), (log(2.)/120.), (log(2.)/600.), .025, 1.]
+gen_jumps!(rs)
```
[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply test/mass_act_jump_tests.jl]: I think it's going to be needed. Doing eval in the same function as the generated function is used will give world age issues, so the user needs two separate calls.

[the remark this answers is not in the document]
Comment (isaacsas) [reply test/mass_act_jump_tests.jl]: I'll wait a bit for @TorkelE or @korsbo to chime in with their thoughts on changing the interface. If no one objects or brings up other issues I'll then update the tests and get this ready for merging.

One other possibility would be to leave the `reaction_network` macro as constructing everything, but then provide a second macro that can be used in the query-based version. The `Union` based `make_type` code could still be used for both, so this basically just amounts to making a second macro function and corresponding simplified `coordinate` function.

[the remark this answers is not in the document]
Comment (TorkelE) [reply test/mass_act_jump_tests.jl]: (Been elsewhere all day but reading through things now. Will have a proper reply within a few hours. Biggest thing is trying to understand the world age problem and why that approach won't work out.)

[the remark this answers is not in the document]
Comment (isaacsas) [reply test/mass_act_jump_tests.jl]: No rush, better to get this right than hurry it! We don't want to have to change the interface again in the near future...

[the remark this answers is not in the document]
Comment (isaacsas) [reply test/mass_act_jump_tests.jl]: In case it helps since it is no longer showing on the test, the world age error arises when using the `DiffEqJump` `Direct` method with `ConstantRateJumps` when the `fill_cur_rates` function is called. This function recursively splats out and evaluates one jump rate at a time until they've all been evaluated. I believe recursion was used instead of looping through the rate functions for performance and type stability reasons. However, this seems to then induce the world age issue since the jump rate functions are newer than the world in which `fill_cur_rates` ends up living.

The bigger issue is that even if we can fix this specific case, I guess there is the potential that (future) code within `ODEProblem`, `SDEProblem` or `JumpProblem` could do something similar, leading to the same error.

**Comment by TorkelE** [comment]
Ok, I think I have understood this world age issue now. The fourth post in this discussion have a good description https://discourse.julialang.org/t/help-calling-a-function-defined-from-expressions/1340/2. You guys probably have a good grip of it already and can skip the following section (or read and correct me if I am wrong). Wrote it down anyway to help myself understand what is going on. (Metaprogramming tends to chew my brain up unless I reprocess what is going on in several ways...)

---

I tried creating a minimum is example corresponding to our case.
```julia
mutable struct SS 
    f::Union{Function,Nothing}
    f_expr::Expr
end
ss = SS(nothing,:(x->2x))
function set_f(zz::SS)
    zz.f = eval(zz.f_expr)
end
set_f(ss)
ss.f(2)
```
We have a mutable structure with a function/nothing field. The field is initially set to nothing. 
We then have a function which sets this field to a real function by evaluating a an expression using the `eval()` function.
At some point we then use that function to calculate the function to have in the field.
We then evaluate the function in the field an gets the number 4.
All of this works fine.

However, here:
```julia
sss = SS(nothing,:(x->2x))
function master_function(zz,x)
    set_f(zz)
    zz.f(x)
end
master_function(sss,2)
```
we still create an object of the structure like before. However, this time we have created some master function that does a lot of stuff for us (in this case both sets the function and evaluates it, but in another example it might be a lot more stuff). This yields a world age error.  When `master_function` is compiled it does not know how the function `zz.f` might look like (since it is only set while the function is being evaluated and eval is done during runtime?). 

---

My fear is that we might get recurring problems with this if we calculate the `.f`, `.g` and the jumps functions separately using `eval()`. Moving that computations out `ODEProblem` etc. might solve some cases of it but more my surface eventually. I wouldn't be surprised if people write their own functions doing stuff which ends up yielding this obscure metaprogramming related world age errors (and lots of confusion for the user with no meta programming experience).

I am not sure whenever this can be entirely avoided though. If we cannot we will have to minimise the instances where this can happen. Is our main concern is to allow the creation of `reaction_networks` with a large number of components, which is then to be investigated in some way before it eventually might be simulated? 

In this case it might be best to allow for a separate macro, e.g. `@sparse_reaction_network` which generates a scaled down reaction network having some of the features but without the heavy calculations. We provide a function which generates a detailed reaction network of the sparse one.

Next we will set up inheritance so that the detailed network inherits from the sparse one (`AbstractReactionNetwork<:SparseAbstractReactionNetwork`, we might have to do some renaming. A problem might be that we are currently using that `AbstractReactionNetwork<:Function`, and making `SparseAbstractReactionNetwork<:Function` might be weird since being a function is the one trait the sparse ones are missing). The inheritance will allow people to define some methods for networks in general, and e.g. `ODEProblem` requires a full one. This still might leave some instances were this might happen, but might allow us to restrict us to a small case.

Finally; This entire issue _might_ be solvable by making the noise matrix and the Jacobian sparse. This would move the computation time from roughly NxN to N, which might actually save a lot of time. In you example, try putting @time in front of the lines
```julia
@time f_expr = get_f(reactions, reactants)
@time g_expr = get_g(reactions, reactants, scale_noise)
@time (jump_rate_expr, jump_affect_expr, jumps, regular_jumps) = get_jumps(reactions, reactants, parameters)
@time symjac = Expr(:quote, calculate_jac(deepcopy(f_rhs), syms))
```
and see the output when you create a large network (I tried this, but I think you have much much larger networks available). If the time from the linear calculation in `@time f_expr = get_f(reactions, reactants)` is at an acceptable level and the long runtime is all due to the quadratic computations we might be able to fix much of the issue simply by using sparse matrices (and even if everything is not solved, we might further reduce the number of large cases where one have to do stuff risking the world age problem).

**Comment by TorkelE** [comment]
(mistakenly pushed "Close and comment" instead of "Comment"...)

**Comment by isaacsas** [comment]
Just to add, I do think eventually switching to a sparse Jacobian for larger systems makes sense for both macros, but I suspect that for sufficiently large systems there will still noticeable (though greatly lessened) overhead in calculating Jacobians, `f` and/or `g` functions (which would be great to eliminate for people that don't need ODE/SDE problems).

**Comment by TorkelE** [comment]
I will have to admit that I am not really sure how to interpret the benchmark, the first line seems to take a lot of time in the first two cases (is this `reactions = get_reactions(ex)`? If that's the case we have problems since that line is not avoidable by moving out the creation of the `f` and `g` functions). The lines 21/22/23, somewhere around which I expect the Jacobian to be calculated, does not take much time and it is here I expect big troubles.

The `Function` subtype is only used when solving ODEs etc. (I think by the entire thing being a function it gets passed on to the solution object. This allows e.g. the `plot()` function to have info about reactant names and automatically add it into figures.).

You are right that the entire package right now is very tailored towards modelling and simulations. We provide very little functionality for people looking at other aspects of these networks (topologies etc.), and if people do these kinds of things certain aspects of the simulation part might cause problem (like the creation time).

Maybe we should have two types of reaction networks `ReactionNetwork` and `ReactionNetworkModel`? It would not be hard to allow for conversion between them (and possible inheritance). Admittedly I am (mostly) used to the simulation parts, but what things do you think is most important to have for people not studying that aspect? (The dependency graph, but we have that one right?)

**Comment by isaacsas** [comment]
My interpretation is that for tens to hundreds of reactions, calling `get_reactions` is the dominant cost. In those cases the macro is pretty fast (a few seconds at most), so I don't think we need to worry about relative performance. Moreover, this cost (and corresponding memory allocation) seems to be relatively fixed -- it is basically the same across all three networks. The last example is a large network with over three thousand reactions. For such networks It can take minutes to even longer load them in, with the dominant costs coming from calculating `g`, `g_expr` and `symjac` (lines 11, 10 and 15).

The automatic plotting with symbols could be resolved by creating and passing an `ODEFunction` or `SDEFunction` in the problem creation. @ChrisRackauckas can hopefully clarify, but my impression is that this is now the standard way to feed information into those types of problems. (In fact, it would be nice to have something similar for jumps, since right now we don't get the automatic legend using `syms` with them...) We could add fields for them within the macro generated type, which are then generated automatically for the full macro, and added by a second function call for the reduced macro.

Regarding network analysis, there was lots of discussion back in [43](https://github.com/JuliaDiffEq/DiffEqBiological.jl/issues/43) of things people might like to do directly from networks. The information in a `ReactionNetworkModel` should be all that is needed for many such analysis. I've already written some code to calculate various types of dependency graphs needed by SSAs, but it might make sense to ultimately factor some of this out into a more concrete interface for querying network properties from `ReactionNetwork`s.

**Comment by isaacsas** [comment]
FWIW, I think one could still setup the the inheritance
`AbstractMinReactionNetwork <: Function`
`AbstractReactionNetwork <: AbstractMinReactionNetwork`
It's just that `MinReactionNetwork` wouldn't actually be usable as a `Function` until `gen_odes!` is called. Note, this change would need to go in `DiffEqBase`, as that is where `AbstractReactionNetwork` is defined.

I'm going to setup a second PR that creates a `min_reaction_network` macro. I'll close this one once I get it submitted.
